### PR TITLE
Try to build the Docker image if it doesn't exist

### PR DESCRIPTION
**Pull request by huydhn**
There is a bug in the test workflow where it could fail to find the new Docker image when the image hasn't yet became available on ECR, for example https://hud.pytorch.org/pytorch/pytorch/commit/e71ab214226af1f9dbded944e939c6447e0e8f09.  This basically is a race condition where the test job starts before the docker-build workflow could finish successfully.  The fix here is to make sure that the test job has the opportunity to build the image if it doesn't exist, same as what the build workflow does atm.  Once the docker-build workflow finishes pushing the new image to ECR, that can then be used instead.

### CODE REVIEW

File: .github/workflows/_docs.yml [hunk 1]
```diff
@@ -81,10 +81,17 @@ jobs:
       - name: Setup Linux
         uses: ./.github/actions/setup-linux
 
+      - name: Calculate docker image
```
Comment (PaliC) [inline .github/workflows/_docs.yml:84]: Why did you move this to _docs?

**Review by PaliC — COMMENTED** [review]
I'm also a bit unsure what the retry mechanism is, can you point me to it?

**Comment by huydhn** [comment]
> I'm also a bit unsure what the retry mechanism is, can you point me to it?

Oh, there is no retry, the change is to add the step that looks for and build Docker image if it doesn’t exist (calculate-docker-image)

File: .github/workflows/_docs.yml (hunk 1 above)
Comment (huydhn) [reply .github/workflows/_docs.yml:84]: I'm adding this step to all occurrences of `pull-docker-image` to make sure that pulling docker image would not fail because it couldn't find the image

**Review by PaliC — APPROVED** [review]
lgtm

**Comment by malfet** [comment]
There is a slight possibility of race condition, but it opens up a window for huge waste of GPU cycles on building those images. For example https://github.com/pytorch/pytorch/pull/102674 is based on https://github.com/pytorch/pytorch/commit/9331b7fa05d92c76e0325a848c50e2756ce6cfaa which was landed 7 hours ago, so it's not a race condition, but rather a docker image is missing from a registry.

**Comment by huydhn** [comment]
> There is a slight possibility of race condition, but it opens up a window for huge waste of GPU cycles on building those images. For example #102674 is based on [9331b7f](https://github.com/pytorch/pytorch/commit/9331b7fa05d92c76e0325a848c50e2756ce6cfaa) which was landed 7 hours ago, so it's not a race condition, but rather a docker image is missing from a registry.

The errors on that step is about an invalid Docker image though `308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-bionic-cuda11.8-cudnn8-py3-gcc7:2449172a5f2d2b623e0baab605db55c4105e35f5:2449172a5f2d2b623e0baab605db55c4105e35f5`.  You can see that the prefix `308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/` appears twice. The GHA `calculate-docker-image` was updated by this PR to handle this case.  It now accepts both image name like `pytorch-linux-bionic-cuda11.8-cudnn8-py3-gcc7` and Docker image name like `308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-bionic-cuda11.8-cudnn8-py3-gcc7:2449172a5f2d2b623e0baab605db55c4105e35f5`.  Before the change, only the former was valid.

What makes me think this is a race condition because the missing Docker image error only popped up when people tried to update the image, i.e. https://hud.pytorch.org/pytorch/pytorch/commit/e71ab214226af1f9dbded944e939c6447e0e8f09.  The build job with `calculate-docker-image` passed while the test jobs without it failed.  Retry then works after the image(s) became available.  Do I miss anything when assuming what only `docker-builds.yml` has the power to push the image to ECR?
